core/arch/arm/pta/core_fs_htree_tests.c

```diff
@@ -0,0 +1,456 @@
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <assert.h>
+#include <string.h>
+#include <tee/fs_htree.h>
+#include <tee/tee_fs_rpc.h>
+#include <trace.h>
+#include <types_ext.h>
+#include <util.h>
+
+#include "core_self_tests.h"
+
+/*
+ * The smallest blocks size that can hold two struct
+ * tee_fs_htree_node_image or two struct tee_fs_htree_image.
+ */
+#define TEST_BLOCK_SIZE		144
+
+struct test_aux {
+	uint8_t *data;
+	size_t data_len;
+	size_t data_alloced;
+	uint8_t *block;
+};
+
+static TEE_Result test_get_offs_size(enum tee_fs_htree_type type, size_t idx,
+				     uint8_t vers, size_t *offs, size_t *size)
+{
+	const size_t node_size = sizeof(struct tee_fs_htree_node_image);
+	const size_t block_nodes = TEST_BLOCK_SIZE / (node_size * 2);
+	size_t pbn;
+	size_t bidx;
+
+	COMPILE_TIME_ASSERT(TEST_BLOCK_SIZE >
+			    sizeof(struct tee_fs_htree_node_image) * 2);
+	COMPILE_TIME_ASSERT(TEST_BLOCK_SIZE >
+			    sizeof(struct tee_fs_htree_image) * 2);
+
+	assert(vers == 0 || vers == 1);
+
+	/*
+	 * File layout
+	 *
+	 * phys block 0:
+	 * tee_fs_htree_image vers 0 @ offs = 0
+	 * tee_fs_htree_image vers 1 @ offs = sizeof(tee_fs_htree_image)
+	 *
+	 * phys block 1:
+	 * tee_fs_htree_node_image 0  vers 0 @ offs = 0
+	 * tee_fs_htree_node_image 0  vers 1 @ offs = node_size
+	 *
+	 * phys block 2:
+	 * data block 0 vers 0
+	 *
+	 * phys block 3:
+	 * tee_fs_htree_node_image 1  vers 0 @ offs = 0
+	 * tee_fs_htree_node_image 1  vers 1 @ offs = node_size
+	 *
+	 * phys block 4:
+	 * data block 0 vers 1
+	 *
+	 * ...
+	 */
+
+	switch (type) {
+	case TEE_FS_HTREE_TYPE_HEAD:
+		*offs = sizeof(struct tee_fs_htree_image) * vers;
+		*size = sizeof(struct tee_fs_htree_image);
+		return TEE_SUCCESS;
+	case TEE_FS_HTREE_TYPE_NODE:
+		pbn = 1 + ((idx / block_nodes) * block_nodes * 2);
+		*offs = pbn * TEST_BLOCK_SIZE +
+			2 * node_size * (idx % block_nodes) +
+			node_size * vers;
+		*size = node_size;
+		return TEE_SUCCESS;
+	case TEE_FS_HTREE_TYPE_BLOCK:
+		bidx = 2 * idx + vers;
+		pbn = 2 + bidx + bidx / (block_nodes * 2 - 1);
+		*offs = pbn * TEST_BLOCK_SIZE;
+		*size = TEST_BLOCK_SIZE;
+		return TEE_SUCCESS;
+	default:
+		return TEE_ERROR_GENERIC;
+	}
+}
+
+static TEE_Result test_read_init(void *aux, struct tee_fs_rpc_operation *op,
+				 enum tee_fs_htree_type type, size_t idx,
+				 uint8_t vers, void **data)
+{
+	TEE_Result res;
+	struct test_aux *a = aux;
+	size_t offs;
+	size_t sz;
+
+	res = test_get_offs_size(type, idx, vers, &offs, &sz);
+	if (!res) {
```

Comment:
Maybe now isn't the time to introduce this style. It it quite analogous to the kernel style otherwise in use. Let's discuss that at some other time. I'll change here.